src/test/unit/model/indexing/rvalue_cl_test.cpp

```diff
@@ -0,0 +1,273 @@
+          expect_eq(rvalue(m, "", ind1), from_matrix_cl_nonscalar(rvalue(
+                                             m_cl, "", opencl_index(ind1))));
+          expect_eq(
+              rvalue(m_i, "", ind1),
+              from_matrix_cl_nonscalar(rvalue(m_i_cl, "", opencl_index(ind1))));
+
+          stan::math::matrix_v m_v1 = m;
+          stan::math::matrix_v m_v2 = m;
+          stan::math::var_value<stan::math::matrix_cl<double>> m_v_cl
+              = stan::math::to_matrix_cl(m_v2);
+          auto correct = rvalue(m_v1, "", ind1);
+          auto res = from_matrix_cl_nonscalar(
+              rvalue(m_v_cl, "", opencl_index(ind1)));
+          expect_eq(correct.val(), res.val());
+          set_adjoints(correct);
+          set_adjoints(res);
+          stan::math::grad();
+          expect_eq(m_v1.adj(), m_v2.adj());
+          stan::math::recover_memory();
+        }(std::get<Is1>(indices)),
+        0)...};
+  });
+}
+
+TEST(ModelIndexing, rvalue_opencl_matrix_2d) {
+  Eigen::MatrixXd m(4, 4);
+  m << 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16;
+  stan::math::matrix_cl<double> m_cl(m);
+  Eigen::MatrixXi m_i(4, 4);
+  m_i << 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16;
+  stan::math::matrix_cl<int> m_i_cl(m_i);
+  auto indices = std::make_tuple(
+      index_uni(3), index_omni(), index_multi(std::vector<int>{1, 2, 1, 3, 1}),
+      index_min(2), index_max(3), index_min_max(2, 3), index_min_max(3, 1));
+  constexpr int N_ind = std::tuple_size<decltype(indices)>::value;
+  stan::math::index_apply<N_ind>([&](auto... Is1) {
+    std::initializer_list<int>{(
+        [&](const auto& ind1) {
+          stan::math::index_apply<N_ind>([&](auto... Is2) {
+            std::initializer_list<int>{(
+                [&](const auto& ind2) {
+                  expect_eq(
+                      rvalue(m, "", ind1, ind2),
+                      from_matrix_cl_nonscalar(rvalue(
+                          m_cl, "", opencl_index(ind1), opencl_index(ind2))));
+                  expect_eq(
+                      rvalue(m_i, "", ind1, ind2),
+                      from_matrix_cl_nonscalar(rvalue(
+                          m_i_cl, "", opencl_index(ind1), opencl_index(ind2))));
+
+                  stan::math::matrix_v m_v1 = m;
+                  stan::math::matrix_v m_v2 = m;
+                  stan::math::var_value<stan::math::matrix_cl<double>> m_v_cl
+                      = stan::math::to_matrix_cl(m_v2);
+                  auto correct = rvalue(m_v1, "", ind1, ind2);
+                  auto res = from_matrix_cl_nonscalar(rvalue(
+                      m_v_cl, "", opencl_index(ind1), opencl_index(ind2)));
+                  expect_eq(correct.val(), res.val());
+                  set_adjoints(correct);
+                  set_adjoints(res);
+                  stan::math::grad();
+                  expect_eq(m_v1.adj(), m_v2.adj());
+                  stan::math::recover_memory();
```

Comment:
Are you testing somewhere that the values are the same after grad() when slicing assignment happens?

Reply:
What do you mean by slicing assignment?

Reply:
So when we have 

```stan
matrix[N, M] x = exp(y);
x[1:10, 1:10] = exp(x);
```

Or some slicing assignment like that, making sure that the values after the forward pass are correct and that they are put back after the reverse pass.

Reply:
Tests for assignments are in `assign_cl_test.cpp`. See for example line 92 in that file.